### ISSUE TITLE
Feature: LTO

### ISSUE THREAD

**Issue by pchome** [azerothcore/azerothcore-wotlk#17014]
### Describe your feature request or suggestion in detail

Currently there are multiple ODR (One Definition Rule) violation warnings with LTO (Link Time Optimization) enabled for POD structures (Plain Old Data: enum, struct, etc.).

I decided to abandon my old MaNGOS build for more active community when I finally in mood to fix this, so I'm currently on the **first time build** stage. Seems many errors are still present at least in `src/server/scripts/*`.
```console
src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingLair/boss_chromaggus.cpp:64:
warning: type ‘Misc’ violates the C++ One Definition Rule [-Wodr]
   64 | enum Misc
      | 
src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingLair/boss_nefarian.cpp:180:
note: an enum with different value name is defined in another translation unit
  180 | enum Misc
      | 
```
_(the output formatted for readability by me)_

### Describe a possible solution to your feature or suggestion in detail

IMHO most simple (and maybe temporary) fix will be just renaming all of them (enums/structs/...) by prefixing some context.
w/o deep diving where it used in code later e.g. for `boss_chromaggus.cpp`:
```cpp
enum Misc
{
    GUID_LEVER_USER = 0
};
```
to
```cpp
enum BChrMisc
{
    GUID_LEVER_USER = 0
};
using Misc = BChrMisc;
```

### Additional context

Maybe I'll PR a fix by myself later, or feel free to outrun me (preferred).

OS: Gentoo
GCC: 12.3.1
CFLAGS: are very similar to ones used by the [gentooLTO](https://github.com/InBetweenNames/gentooLTO) project:
`export CFLAGS='-march=native -O3 -pipe -fgraphite-identity -ftree-loop-distribution -floop-nest-optimize -flto=auto -fuse-linker-plugin -fdevirtualize-at-ltrans -Wl,-O1 -Wl,--as-needed -Wl,--hash-style=gnu'`
`export CXXFLAGS="$CFLAGS"`
`export LDFLAGS="$CFLAGS"`

**Comment by Nyeriah** [comment]
This is the trinitycore scripting convention

https://github.com/TrinityCore/TrinityCore/blob/master/src/server/scripts/Northrend/IcecrownCitadel/boss_the_lich_king.cpp
https://github.com/TrinityCore/TrinityCore/blob/master/src/server/scripts/Northrend/IcecrownCitadel/boss_festergut.cpp

etc.

The convention is used throughout the entire repository

**Comment by Kitzunu** [comment]
imo I don't see a reason to change it. It gives consistency between scripts while not breaking anything.

My suggestion to you is to escape the warning if it makes compiling on your OS fail, as we don't support Gentoo

**Comment by walkline** [comment]
@pchome Could you explain to a person who knows almost nothing about C/C++ compilers what benefits the LTO feature can bring?

**Comment by pchome** [comment]
@Kitzunu 
> My suggestion to you is to escape the warning if it makes compiling on your OS fail, as we don't support Gentoo

It compiles. But from my experience from MaNGOS it will fail the server randomly. So I ended up disabling LTO for mangos.
Now I just trying reuse the same bash script to build azurecore also relaxing some flags.

> ... we don't support Gentoo

_an distribution focused on building stuff from source ..., I'm just building from source then on random distro ;)_ You'd better support this for ... code quality, idk.

So give me some guide, what's next? We'll live this issue open (and abandoned, aka GTFO), _wontfix_ or some PRs are welcomed?
Either way it is not a big deal for me too, I'll continue experimenting just disabling some troubling flags.
But this issue may bite you in the future.

**Comment by walkline** [comment]
Sounds reasonable to me. And how many of these kinds of warnings do you have? Do you see any other solutions besides renaming object definitions?

**Comment by pchome** [comment]
https://en.cppreference.com/w/cpp/language/definition#One_Definition_Rule

There are at least dozen of such warning, but
according to my experience they should be all renamed, even wasn't warned about. I just renamed the first for testing and got a bunch of new warnings in the other script sources.

IMHO to be sure: just (imaginary) place some of two scripts in one file, see conflicting names. Welcome to the [unity builds](https://en.wikipedia.org/wiki/Unity_build) ;)

**Comment by walkline** [comment]
@Kitzunu @Nyeriah What are your thoughts regarding this? Perhaps we could change the naming convention in the scripts?

**Comment by pchome** [comment]
It could be a regex game for me (or people), e.g. I believe I could rename all of them with just one linux console command ... I just waiting for final decision.

**Comment by Nyeriah** [comment]
> @Kitzunu @Nyeriah What are your thoughts regarding this? Perhaps we could change the naming convention in the scripts?

I don’t see a reason to change the convention at this time and I will echo what @Kitzunu has already said about this

**Comment by Kitzunu** [comment]
I get where you're coming from, as you seem quite invested and interested in it. I got stuff myself I am interested in but get turned down. It's a bummer but sometimes it is what it is.

I also assume the x10 shrinkage of compilation time is a genral number and not tested for AC. And to be quite frank I'd say compilation times in general are not that big of an issue on AC to the point where we need to go change the naming convention of every script we use.

I am all for general optimization, but in this case I do think that changing the naming convention is just not neccessary for the, what I would belive to be, small gain in compilation times.

**Comment by michaeldelago** [comment]
I personally don't care one way or another; If there's a performance benefit to changing the convention in the scripts, we can change the convention. Additionally, while I understand AC's lineage with Trinity, there's no hard reason to follow their conventions, right?

I lean towards the @pchome's issue, I think it's better if we can abide by stricter linting/warning levels. I use [GNU Guix](https://guix.gnu.org/) and that's similar to Gentoo in that it's a source based Linux distro[0].

For those who may not be wholly aware, LTO isn't about compilation times. It's about performance during execution. Here's some benchmarks for the Linux Kernel with and without LTO

https://www.phoronix.com/review/clang-lto-kernel

The TL;DR is that Clang with LTO had an average 9% performance boost over GCC without LTO.

High level, I don't know if we'd have the same improvement in AzerothCore, and I have no idea on how I could reasonably stress test this. 

I find that large changes like this are much easier if you do the legwork. Make the changes to the scripts, confirm it builds, try and see if you can accurately benchmark CPU usage. My best guess would be to measure CPU percentage while doing something specific before and after applying the changes to the scripts, though I don't know how well that'd show a performance change. 

[0] - I'm actually working on [a build/package for AC](https://github.com/michaeldelago/delaguix/blob/main/delaguix/packages/azerothcore.scm#L40). Biggest thing in my way (which I'm also working on 🙃) is modules as shared libraries

**Comment by Kitzunu** [comment]
No, there isnt a reason to stick to TC conventions, but for me to even consider changing the entire naming convention I'd want to see real results on AC from it.

Also I would like to know what benefits it could bring to msvc builds

**Comment by pchome** [comment]
@Kitzunu 
> Also I would like to know what benefits it could bring to msvc builds

LTO named differently in msvc ... `/LTCG` (Link-time code generation) (**?**)

No need to support/hardcode this in your build system, just make sure everithing builds ok when user enables this flag(s).

**BTW**: building auth server currently ok with LTO, but not world server.
If you have some intensive tests for auth server then ...

**Moreover**: if user decides to use PGO/FDO/AFDO/Bolt/... for server then benefit from +LTO should be even noticeable on 'busy' servers (internet said this to me).

FYI, because I too don't care about this in foreseeable future.
